crates/polars-io/src/hive.rs

```diff
@@ -54,35 +54,52 @@ pub(crate) fn materialize_hive_partitions<D>(
     }
 }
 
-/// Merge 2 sets of columns into one, where each set contains columns ordered such that their indices
+/// Merge 2 lists of columns into one, where each list contains columns ordered such that their indices
 /// in the `schema` are in ascending order.
 ///
 /// Layouts:
-/// * `df_columns`: `[row_index?, ..schema_columns]`
-///   * `df_columns` must start with either a row_index column, or a schema column. This is important
-///     as we assume that the first column in `df_columns` is a row_index column if it doesn't exist
-///     in the `schema`.
-/// * `hive_columns`: `[..schema_columns?, ..hive_columns?]`
+/// * `cols_lhs`: `[row_index?, ..schema_columns?, ..other_left?]`
+///   * If the first item in `cols_lhs` is not found in the schema, it will be assumed to be a
+///     `row_index` column and placed first into the result.
+/// * `cols_rhs`: `[..schema_columns? ..other_right?]`
+///
+/// Output:
+/// * `[..schema_columns?, ..other_left?, ..other_right?]`
+///
+/// Note: The `row_index` column should be handled before calling this function.
 ///
 /// # Panics
-/// Panics if either `df_columns` or `hive_columns` is empty.
+/// Panics if either `cols_lhs` or `cols_rhs` is empty.
 pub fn merge_sorted_to_schema_order<'a, D>(
-    df_columns: &'a mut dyn ExactSizeIterator<Item = Column>,
-    hive_columns: &'a mut dyn ExactSizeIterator<Item = Column>,
+    cols_lhs: &'a mut dyn Iterator<Item = Column>,
+    cols_rhs: &'a mut dyn Iterator<Item = Column>,
     schema: &polars_schema::Schema<D>,
     output: &'a mut Vec<Column>,
 ) {
-    // Safety: Both `df_columns` and `hive_columns` are non-empty.
-    let mut series_arr = [df_columns.peekable(), hive_columns.peekable()];
-
-    if let Some(i) = schema.index_of(series_arr[1].peek().unwrap().name()) {
-        let mut schema_idx_arr = [
-            // `unwrap_or(0)`: The first column could be a row_index column that doesn't exist in the `schema`.
-            schema
-                .index_of(series_arr[0].peek().unwrap().name())
-                .unwrap_or(0),
-            i,
-        ];
+    merge_sorted_to_schema_order_impl(cols_lhs, cols_rhs, output, &|v| schema.index_of(v.name()))
+}
+
+pub fn merge_sorted_to_schema_order_impl<'a, T, O>(
```

Comment:
This generic `_impl` gets used by projection pushdown for schema merging, where `T` becomes the schema fields.